src/lib/virtio/net_driver.lua

```diff
@@ -0,0 +1,171 @@
+-- Application to connect to a virtio-net driver implementation
+--
+-- Licensed under the Apache 2.0 license
+-- http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Copyright (c) 2015 Virtual Open Systems
+--
+
+module(..., package.seeall)
+
+local debug = _G.developer_debug
+
+local ffi       = require("ffi")
+local C         = ffi.C
+local S         = require('syscall')
+local pci       = require("lib.hardware.pci")
+local lib       = require("core.lib")
+local bit       = require('bit')
+local virtq     = require('lib.virtio.virtq_driver')
+local VirtioPci = require('lib.virtio.virtio_pci').VirtioPci
+local checksum  = require('lib.checksum')
+require('lib.virtio.virtio_h')
+
+local band, bor, rshift, lshift = bit.band, bit.bor, bit.rshift, bit.lshift
+local prepare_packet4, prepare_packet6 = checksum.prepare_packet4, checksum.prepare_packet6
+local new_packet = packet.allocate
+
+-- constants
+local ETHERTYPE_IPv4 = C.htons(0x0800)
+local ETHERTYPE_IPv6 = C.htons(0x86DD)
+local ETHERTYPE_OFF = 12
+local ETHERLEN = 14 -- DST MAC | SRC MAC | ethertype
+local VIRTIO_NET_HDR_F_NEEDS_CSUM = 1
+
+local min_features =   C.VIRTIO_RING_F_INDIRECT_DESC + C.VIRTIO_NET_F_CSUM
+local want_features =  C.VIRTIO_F_ANY_LAYOUT +
+                       C.VIRTIO_RING_F_INDIRECT_DESC +
+                       C.VIRTIO_NET_F_CSUM +
+                       C.VIRTIO_NET_F_MAC
+
+local RXQ = 0
+local TXQ = 1
+
+VirtioNetDriver = {}
+VirtioNetDriver.__index = VirtioNetDriver
+
+function VirtioNetDriver:new(args)
+
+   local virtio_pci = VirtioPci:new(args.pciaddr)
+
+   virtio_pci:reset()
+   virtio_pci:acknowledge()
+
+   virtio_pci:driver()
+
+   local error = virtio_pci:set_guest_features(min_features, want_features)
+   if error then
+      virtio_pci:free()
+      return nil, error
+   end
+
+   if debug then print("enumerating queues...") end
+   local vqs = {}
```

Comment:
seems that if this table starts from 0, it should be an ffi table; dunno, just a thought